File: 15/cpp/main.cpp
#include <iostream>
#include <set>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        std::set<std::set<int>> triplets;
        if (nums.size() < 3) { return {}; }

        // Three flags
        int right = 2;
        for (int left = 0; left + 2 < nums.size(); left++) {
            for (int middle = left + 1; middle + 1 < nums.size(); middle++) {
                for (int right = middle + 1; right < nums.size(); right++) {
                    if (nums[left] + nums[middle] + nums[right] == 0) {
                        triplets.insert({nums[left], nums[middle], nums[right]});
                    }
                }
            }
        }

        // Convert set of sets to vector of vectors
        std::vector<std::vector<int>> result;
        for (const auto& triplet : triplets) {
            result.push_back(std::vector<int>(triplet.begin(), triplet.end()));
        }

        return result;
    }
};

int main() {
    Solution sol;

    // Test case X: Y
    RETURN_TYPE resultX = sol.FUNCTION_NAME(Y);
    std::cout << "Test X: " << resultX << std::endl;

    return 0;
}

File: 15/python3/main.py
#!/usr/bin/env python3

class Solution:
    def threeSum(self, nums):
        # Sort the input list
        nums.sort()
        result = []
        
        for i in range(len(nums)):
            # Skip duplicate elements to avoid duplicate triplets
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                
                if total == 0:
                    # Add the triplet to the result
                    result.append([nums[i], nums[left], nums[right]])

                    # Skip duplicates for both left and right pointers
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1

                    # Move both pointers inward
                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1

        return result
